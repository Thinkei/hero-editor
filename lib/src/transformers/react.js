import React from 'react';
import { Text, Node } from 'slate';
import y from '../helpers/why';

const isRoot = Array.isArray;

const makeReactTransformer = (rules) =>
  y((reactTransformer) => (node, key) => {
    if (isRoot(node)) return node.map(reactTransformer);

    if (Text.isText(node)) return rules.leaf(node, key);

    const children = node.children.map(reactTransformer);

    return rules.element(node, children, key);
  });

const defaultRules = {
  leaf: (node, key) => {
    let children = node.text;

    if (node.bold) children = <strong key={key}>{children}</strong>;

    if (node.italic) children = <em key={key}>{children}</em>;

    if (node.underline) children = <u key={key}>{children}</u>;

    return <span key={key}>{children}</span>;
  },

  element: (node, children, key) => {
    switch (node.type) {
      case 'paragraph':
        return (
          <p
            key={key}
            style={{ margin: 0, lineHeight: 1.5, minHeight: '1.5em' }}
          >
            {children}
          </p>
        );

      case 'heading-one':
        return <h1 key={key}>{children}</h1>;

      case 'link':
        return (
          <a key={key} href={Node.string(node)} target="_blank">
            {children}
          </a>
        );

      case 'mention':
        return (
          <span
            key={key}
            style={{
              padding: '2px 4px',
              background: '#eeeeee',
              borderRadius: 4,
            }}
          >
            @{node.data.name}
          </span>
        );

      case 'bulleted-list':
        return <ul key={key}>{children}</ul>;

      case 'numbered-list':
        return <ol key={key}>{children}</ol>;

      case 'list-item':
        return <li key={key}>{children}</li>;

      default:
        return <div key={key}>{children}</div>;
    }
  },
};

const defaultReactTransformer = makeReactTransformer(defaultRules);

export default makeReactTransformer;

export { makeReactTransformer, defaultReactTransformer };
