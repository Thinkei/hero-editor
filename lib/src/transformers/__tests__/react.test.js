import { render } from '@testing-library/react';
import { defaultReactTransformer } from '../react';

describe('reactTransformer', () => {
  it('works properly', () => {
    const { baseElement } = render(defaultReactTransformer(value));
    expect(baseElement).toMatchSnapshot();
  });
});

const value = [
  {
    type: 'paragraph',
    children: [
      {
        text: 'A',
        bold: true,
        underline: true,
      },
      {
        text: ' ',
        underline: true,
      },
      {
        text: 'paragraph',
        italic: true,
        underline: true,
      },
    ],
  },
  {
    type: 'paragraph',
    children: [
      {
        text: 'This is a mention ',
      },
      {
        type: 'mention',
        data: {
          id: 1,
          name: 'Toan Nguyen',
        },
        children: [
          {
            text: '',
          },
        ],
      },
    ],
  },
  {
    type: 'paragraph',
    children: [
      {
        text: '',
      },
    ],
  },
  {
    type: 'bulleted-list',
    children: [
      {
        type: 'list-item',
        children: [
          {
            text: 'Item 1',
          },
        ],
      },
      {
        type: 'list-item',
        children: [
          {
            text: 'Item 2',
          },
        ],
      },
    ],
  },
  {
    type: 'numbered-list',
    children: [
      {
        type: 'list-item',
        children: [
          {
            text: 'Item 1',
          },
        ],
      },
      {
        type: 'list-item',
        children: [
          {
            text: 'Item 2',
          },
        ],
      },
    ],
  },
];
