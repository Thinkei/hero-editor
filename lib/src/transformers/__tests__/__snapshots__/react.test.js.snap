// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`reactTransformer works properly 1`] = `
<body>
  <div>
    <p
      style="margin: 0px; line-height: 1.5; min-height: 1.5em;"
    >
      <span>
        <u>
          <strong>
            A
          </strong>
        </u>
      </span>
      <span>
        <u>
           
        </u>
      </span>
      <span>
        <u>
          <em>
            paragraph
          </em>
        </u>
      </span>
    </p>
    <p
      style="margin: 0px; line-height: 1.5; min-height: 1.5em;"
    >
      <span>
        This is a mention 
      </span>
      <span
        style="padding: 2px 4px; background: rgb(238, 238, 238); border-radius: 4px;"
      >
        @
        Toan Nguyen
      </span>
    </p>
    <p
      style="margin: 0px; line-height: 1.5; min-height: 1.5em;"
    >
      <span />
    </p>
    <ul>
      <li>
        <span>
          Item 1
        </span>
      </li>
      <li>
        <span>
          Item 2
        </span>
      </li>
    </ul>
    <ol>
      <li>
        <span>
          Item 1
        </span>
      </li>
      <li>
        <span>
          Item 2
        </span>
      </li>
    </ol>
  </div>
</body>
`;
